tests/tablespaces/tablespace_utils.py

```diff
@@ -0,0 +1,113 @@
+import time
+import datetime as dt
+import psycopg2
+
+
+class PGNodeTS:
+    def __init__(
+        self,
+        port,
+        service_name,
+        monitor_node=None,
+    ):
+        self.port = port
+        self.service_name = service_name
+        self.monitor = monitor_node
+        if monitor_node is None:
+            self.monitor = self
+
+    def run_sql_query(self, query, *args):
+        """
+        Runs the given sql query with the given arguments in this postgres node
+        and returns the results. Returns None if there are no results to fetch.
+        """
+        result = None
+        conn = psycopg2.connect(self.connection_string())
+        conn.autocommit = True
+
+        with conn:
+            with conn.cursor() as cur:
+                cur.execute(query, args)
+                try:
+                    result = cur.fetchall()
+                except psycopg2.ProgrammingError:
+                    pass
+        # leaving contexts closes the cursor, however
+        # leaving contexts doesn't close the connection
+        conn.close()
+
+        return result
+
+    def connection_string(self):
+        if self.service_name == "monitor":
+            return (
+                "postgresql://autoctl_node@"
+                + self.service_name
+                + ":"
+                + self.port
+                + "/pg_auto_failover"
+            )
+        else:
+            return (
+                "postgresql://docker@"
+                + self.service_name
+                + ":"
+                + self.port
+                + "/postgres"
+            )
+
+    def get_state(self):
+        results = self.monitor.run_sql_query(
+            "SELECT reportedstate FROM pgautofailover.node WHERE nodeport = '"
+            + self.port
+            + "'"
+        )
+        result = ""
+        if results:
+            result = results[-1]
+            if result:
+                result = result[-1]
+
+        return result
+
+    def wait_until_state(self, target_state):
+        prev_state = None
+        wait_until = dt.datetime.now() + dt.timedelta(seconds=90)
+        while wait_until > dt.datetime.now():
+            time.sleep(3)
+
+            current_state = self.get_state()
+
+            if current_state != prev_state:
+                if current_state == target_state:
+                    print(
+                        "state of %s is '%s', done waiting"
+                        % (self.service_name, current_state)
+                    )
+                else:
+                    print(
+                        "state of %s is '%s', waiting for '%s' ..."
+                        % (self.service_name, current_state, target_state)
+                    )
+
+            if current_state == target_state:
+                return True
+
+            prev_state = current_state
+        raise Exception("never got to state")
```

Comment:
Again, that looks very familiar. Is it from PGNode? How can we re-use existing code instead of duplicating it?